memory/iter.go

```diff
@@ -7,210 +7,174 @@ import (
+	if len(end) > 0 && reverse {
+		iter.Seek(end)
+	} else if len(start) > 0 && !reverse {
+		iter.Seek(start)
+	} else if hasItem {
 		iter.loadLatestItem()
-	} else {
-		iter.curItem = nil
+
+		if !iter.Valid() {
+			iter.Next()
+		}
 	}
+
+	return iter
 }
 
-func (iter *dsPrefixIter) Close(ctx context.Context) error {
-	return iter.it.Close()
+func (iter *iterator) Domain() (start []byte, end []byte) {
+	return iter.start, iter.end
 }
 
-func (iter *dsPrefixIter) loadLatestItem() {
-	curItem := iter.it.Item()
-	for iter.it.Next() {
-		if bytes.Equal(curItem.key, iter.it.Item().key) {
-			curItem = iter.it.Item()
-			continue
-		}
-		iter.it.Prev()
-		break
+func (iter *iterator) Valid() bool {
+	if !iter.hasItem {
+		return false
 	}
 
-	if curItem.isDeleted {
-		iter.curItem = nil
+	if iter.it.Item().isDeleted {
+		return false
+	}
 
-		if iter.it.Next() {
-			iter.loadLatestItem()
-		}
-		return
+	if !iter.mayExactlyMatchStart && (!iter.reverse && bytes.Equal(iter.it.Item().key, iter.start) ||
+		iter.reverse && bytes.Equal(iter.it.Item().key, iter.end)) {
+		return false
 	}
 
-	if len(curItem.key) == 0 {
-		// If the current item doesn't exist, explicitly set the current item to nil
-		// instead of a default value, this saves us from having to check the length
-		// of the `curItem.key` property everywhere.
-		iter.curItem = nil
-		return
+	if len(iter.end) > 0 && !lt(iter.it.Item().key, iter.end) {
+		return false
 	}
 
-	iter.curItem = &curItem
+	return gte(iter.it.Item().key, iter.start)
 }
 
-type baseIterator struct {
-	it btree.IterG[dsItem]
-
-	// The smallest key that is valid (inclusive bound) for this iterator to return
-	lowerBound []byte
-
-	// The smallest key that is invalid (exclusive bound) for this iterator to return.
-	upperBound []byte
+func (iter *iterator) Next() {
+	if !iter.hasItem {
+		return
+	}
 
-	// If true, this iterator should be yielding items in reverse order.
-	reverse bool
-}
+	previousItem := iter.it.Item()
+	iter.hasItem = false
 
-func newRangeIter(ctx context.Context, db *Datastore, start, end []byte, reverse bool, version uint64) *dsRangeIter {
-	rIter := &dsRangeIter{
-		ctx:     ctx,
-		db:      db,
-		version: version,
-		it:      newBaseIterator(db.values, reverse, start, end),
-		start:   start,
-		end:     end,
-		reverse: reverse,
+	for iter.next() {
+		// Scan through until we reach the next key.
+		// It doesn't matter if it is deleted or not.
+		if !bytes.Equal(previousItem.key, iter.it.Item().key) {
+			iter.hasItem = true
+			break
+		}
 	}
 
-	if len(end) > 0 && reverse {
-		rIter.Seek(end)
-	} else if len(start) > 0 && !reverse {
-		rIter.Seek(start)
+	if !iter.hasItem {
+		return
 	}
 
-	rIter.loadLatestItem()
-
-	return rIter
-}
+	iter.loadLatestItem()
 
-func (iter *dsRangeIter) Domain() (start []byte, end []byte) {
-	return iter.start, iter.end
-}
-
-func (iter *dsRangeIter) Valid() bool {
-	if iter.curItem == nil {
-		return false
+	if iter.it.Item().isDeleted {
+		iter.Next()
```

Comment:
question: Another `Next` call is called is the latest item in the initial `Next` call was deleted? This means that if there are many deleted items, one next call will skip all of them? Just confirming if that is the desired behavior? Is there a way for caller to control this logic (not skip and decide themselves when they want to call `Next()` next.

Reply:
Yes.  The stores never returned deleted items, the fact that deleted items exist in an internal implementation detail of this type.